sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -669,11 +672,42 @@ object InferFiltersFromConstraints extends Rule[LogicalPlan] with PredicateHelpe
       val newConditionOpt = conditionOpt match {
         case Some(condition) =>
           val newFilters = additionalConstraints -- splitConjunctivePredicates(condition)
-          if (newFilters.nonEmpty) Option(And(newFilters.reduce(And), condition)) else None
+          if (newFilters.nonEmpty) Option(And(newFilters.reduce(And), condition)) else conditionOpt
```

Comment:
Can we keep this unchanged? We just `conditionOpt` in line 681, 685, 693, and 697?

Reply:
So what would be the benefit of keeping that unchanged? To me, it would make the code look confusing. And in theory the two parts (1. infer `newConditionOpt`; 2. infer `newLeftOp` or `newRightOpt`) will be unsynchronized, leaving part 2 always one iteration behind part 1.